### ISSUE TITLE
Question about hidden state in decoder

### ISSUE THREAD

**Issue by Han8931** [A-Jacobson/minimal-nmt#2]
Thank you for your nice tutorial.

I just have a few questions.

````Python3
class Decoder(nn.Module):
    def __init__(self, target_vocab_size, embed_dim, hidden_dim,
                 n_layers, dropout):
        super(Decoder, self).__init__()
        self.n_layers = n_layers
        self.embed = nn.Embedding(target_vocab_size, embed_dim, padding_idx=1)
        self.attention = LuongAttention(hidden_dim)
        self.gru = nn.GRU(embed_dim + hidden_dim, hidden_dim, n_layers,
                          dropout=dropout)
        self.out = nn.Linear(hidden_dim * 2, target_vocab_size)

    def forward(self, output, encoder_out, decoder_hidden):
        """
        decodes one output frame
        """
        embedded = self.embed(output)  # (1, batch, embed_dim)
        context, mask = self.attention(decoder_hidden[:-1], encoder_out)  # 1, 1, 50 (seq, batch, hidden_dim)
        rnn_output, decoder_hidden = self.gru(torch.cat([embedded, context], dim=2),
                                              decoder_hidden)
        output = self.out(torch.cat([rnn_output, context], 2))
        return output, decoder_hidden, mask
````

1)
In your decoder implementation, you set the input dimension size of `self.gru` as embed_dim+hidden_dim and feed embedded input and context together, but I am not so sure whether this is the Luong's method in the oroginal paper.

2) Why you feed `decoder_hidden[:-1]` into the `self.attention`? As I understand, `decoder_hidden[:-1]` indexes the output except the uppermost layer right?

**Comment by A-Jacobson** [comment]
`decoder_hidden` isn't the rnn output, it's the cell state. the indexing is to select only the last gru's state. pytorch `nn.GRU` has a `n_layers` arg which lets you initialize a stacked gru. If `n_layers` > 1 we want to make sure we only take the state of the last layer.

**Comment by Han8931** [comment]
Thanks for your quick answer, but if you want to use the last layer's state then why did you index like `decoder_hidden[:-1]`. I think it should be `decoder_hidden[-1]` or `decoder_hidden[-1:]`.

**Comment by A-Jacobson** [comment]
You are correct, it should be `decoder_hidden[-1:]`. I suppose I never caught that type since I didn't test with more than two layers in the decoder. `decoder_hidden[-1]` probably won't work since that will remove dimensions of size 1.

**Comment by Han8931** [comment]
Thanks! and I want to ask you one more thing. I think the concatenation of target embedding and context (`rnn_output, decoder_hidden = self.gru(torch.cat([embedded, context], dim=2), decoder_hidden)`) is not mentioned in the original paper. I found out many variations of these kinds of implementation. I think this becomes a kind of convention to improve the performance of NMT. However, I can't understand the motivation behind that. Some people seem to call it 'input feeding'. Could you explain it to me?

**Comment by A-Jacobson** [comment]
It is input feeding, and you are correct in that there are tons of variations of attention out there. I think that's because the original paper isn't that clear, at least it wasn't to me before I dug through the open nmt code and just tried variations until I found the one that worked best for this case. 

As for the motivation, my understanding of that line..in english would be: "given `what we have currently translated` and `the important bits of the original sentence` output the `next word`". the embedded input would be our representation of `what we have currently translated`.

**Comment by A-Jacobson** [comment]
pushed a fix for the decoder mistake and updated to 4.0. let me know if everything works out if you end up training with it. 

[commit](https://github.com/A-Jacobson/minimal-nmt/commit/6893496a22e73bfa4772e711207ce180c685a1be)
